File: guest-examples/csv_processor.py
#!/usr/bin/env python3
"""CSV processor example - reads from mounted file, outputs JSON to stdout.

NOTE: This script requires a Nanvix fix to work properly.
Currently, file mounts don't work when using pre-built FAT images
(needed for Python stdlib). See hyperlight-nanvix README for details.
"""

import csv
import json
import sys


def process_csv(input_path):
    """Read CSV, process data, output JSON result to stdout."""
    try:
        # Read CSV file (mounted via file_mounts)
        with open(input_path, 'r') as csvfile:
            reader = csv.DictReader(csvfile)
            rows = list(reader)

        # Log to stderr
        print(f"Read {len(rows)} rows from {input_path}", file=sys.stderr)

        # Process data (example: count by category, calculate stats)
        result = {
            "success": True,
            "total_rows": len(rows),
            "columns": list(rows[0].keys()) if rows else [],
            "data": rows
        }

        # Output JSON to stdout
        print(json.dumps(result))
        return True

    except Exception as e:
        # Error output as JSON
        print(json.dumps({"success": False, "error": str(e)}))
        return False


if __name__ == "__main__":
    # Input: mounted file at /tmp/test.csv
    success = process_csv("/tmp/test.csv")
    sys.exit(0 if success else 1)

File: examples/csv_processor.rs
//! CSV Processor Example
//!
//! This example demonstrates how to:
//! 1. Mount a file from the host into the guest VM
//! 2. Run a Python script that processes the file
//!
//! NOTE: This example currently requires a Nanvix fix to work.
//! See doc/fix-prebuilt-fat-with-mounts.md for details.
//! When using pre-built FAT images (for Python stdlib), file mounts
//! and ramfs are not currently supported.
//!
//! Run with: cargo run --example csv_processor

use anyhow::Result;
use hyperlight_nanvix::{RuntimeConfig, Sandbox};
use std::fs;

#[tokio::main]
async fn main() -> Result<()> {
    // Setup: Create sample CSV in a temp directory
    let temp_dir = std::env::temp_dir().join("hyperlight-csv-example");
    fs::create_dir_all(&temp_dir)?;
    let input_csv = temp_dir.join("input.csv");

    let csv_content = "name,age,city\n\
                       Alice,30,NYC\n\
                       Bob,25,LA\n\
                       Charlie,35,Chicago\n";
    fs::write(&input_csv, csv_content)?;

    println!("Created sample CSV at: {}", input_csv.display());

    // Get nanvix registry path from environment or use default (dist/ folder)
    let nanvix_registry = std::env::var("NANVIX_REGISTRY").unwrap_or_else(|_| {
        // Default to dist/ folder relative to the project root
        let manifest_dir = env!("CARGO_MANIFEST_DIR");
        format!("{}/dist", manifest_dir)
    });

    // Configure runtime with file mounts
    // Nanvix copies files to FAT automatically
    let config = RuntimeConfig::new()
        .with_log_directory("/tmp/hyperlight-nanvix")
        .with_tmp_directory("/tmp/hyperlight-nanvix")
        .with_nanvix_registry(&nanvix_registry)
        .with_file_mount(
            input_csv.to_string_lossy().to_string(),
            "/tmp/test.csv".to_string(), // Guest path
        );

    let mut sandbox = Sandbox::new(config)?;

    // Run Python CSV processor
    println!("Running CSV processor...");
    sandbox.run("guest-examples/csv_processor.py").await?;

    // Output is printed directly to stdout by the guest
    println!("Processing complete!");

    // Cleanup temp file
    let _ = fs::remove_dir_all(&temp_dir);

    Ok(())
}
